src/Raven.Server/Documents/Sharding/Queries/Suggestions/ShardedSuggestionQueryOperation.cs

```diff
@@ -0,0 +1,185 @@
+                }
+
+                BlittableJsonReaderArray popularityMetadata = null;
+
+                if (suggestionJson.TryGet(Constants.Documents.Metadata.Key, out BlittableJsonReaderObject metadata))
+                    metadata.TryGet(Constants.Documents.Metadata.SuggestionPopularityFields, out popularityMetadata);
+
+
+                if (suggestions.TryGetValue(fieldName, out AggregatedSuggestions aggregatedSuggestion) == false)
+                {
+                    aggregatedSuggestion = new AggregatedSuggestions();
+                    suggestions[fieldName] = aggregatedSuggestion;
+                }
+                else
+                {
+                    aggregatedSuggestion = suggestions[fieldName];
+                }
+
+                aggregatedSuggestion.Suggestions.AddRange(suggestionResult.Suggestions);
+
+                if (popularityMetadata is { Length: > 0 })
+                    AddPopularity(popularityMetadata, aggregatedSuggestion, suggestionResult);
+            }
+        }
+
+        var suggestionOptions = _query.Metadata.SelectFields.Where(x => x.IsSuggest)
+                                                                        .Cast<SuggestionField>()
+                                                                        .Where(x => x.HasOptions)
+                                                                        .ToDictionary(x => x.Name.Value, x => x);
+
+        result.Results = new List<SuggestionResult>(suggestions.Count);
+
+        foreach (var suggestion in suggestions)
+        {
+            var fieldName = suggestion.Key;
+
+            IEnumerable<string> aggregatedSuggestions;
+
+            if (suggestion.Value.SuggestionsWithPopularity is null)
+                aggregatedSuggestions = suggestion.Value.Suggestions;
+            else
+                aggregatedSuggestions = suggestion.Value.SuggestionsWithPopularity.Values.OrderByDescending(suggestWord => suggestWord).Select(suggestWord => suggestWord.Term).ToArray();
+
+            var fieldResult = new SuggestionResult { Name = fieldName };
+
+            if (suggestionOptions.TryGetValue(fieldName, out var field) == false)
+                fieldResult.Suggestions = aggregatedSuggestions.ToList();
+            else
+            {
+                var options = field.GetOptions(_context, _query.QueryParameters);
+
+                fieldResult.Suggestions = aggregatedSuggestions.Take(options.PageSize).ToList();
+            }
+
+            result.Results.Add(fieldResult);
+        }
+
+        return result;
+
+        void AddPopularity(BlittableJsonReaderArray popularityArray, AggregatedSuggestions aggregatedSuggestion, SuggestionResult suggestionResult)
+        {
+            var i = 0;
+
+            aggregatedSuggestion.SuggestionsWithPopularity ??= new Dictionary<string, SuggestWord>();
+
+            foreach (BlittableJsonReaderObject pJson in popularityArray.Cast<BlittableJsonReaderObject>())
```

Comment:
maybe we can deserialize that in one shot?

Reply:
Handled in https://github.com/ravendb/ravendb/pull/15718